File: average-of-levels-in-binary-tree/average-of-levels-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> result; double temp=0; int count=0;
        if(root==NULL) return result;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty())
        {
            temp=0;
            int size=q.size();
            for(int i=0;i<size;i++)
            {
                TreeNode* curr_node=q.front(); q.pop();
                temp+=curr_node->val;
                if(curr_node->left!=NULL) q.push(curr_node->left);    
                if(curr_node->right!=NULL) q.push(curr_node->right);
            }
            temp/=size;
            result.push_back(temp);
        }
        return result;
    }
};

File: average-of-levels-in-binary-tree/average-of-levels-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Double> result=new ArrayList();
    Queue<TreeNode> q=new LinkedList();
    public void bfs(TreeNode root){
        q.add(root);
        while(q.isEmpty()==false){
            double size=q.size(); double sum=0;
            for(double i=0;i<size;i++){
                TreeNode temp=q.remove(); sum+=temp.val;
                if(temp.left!=null) q.add(temp.left);
                if(temp.right!=null) q.add(temp.right);
            }
            sum/=size; result.add(sum);
        }
    }
    public List<Double> averageOfLevels(TreeNode root) {
        bfs(root);
        return result;
    }
}